orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc

```diff
@@ -109,8 +114,20 @@ void unregister_grad_fn(size_t ctx_address)
   PyNodeSharedPointerPool::GetInstance().UnRegisterGradFunc(ctx_address);
 }
 
+// Supposed to be cleared on python program exit or before every forward run to resolve following bug:
+// When training program exits, PyNodeSharedPointerPool destructor is called, if grad_fns_ is not empty,
```

Comment:
It seems like the `PyNodeSharedPointerPool` destructor is private. Where and how does it get called?
Who is invoking `release_variables`? Are we able to somehow use some signal to automatically clear `grad_fns_` when `release_variables` is called?

Reply:
PyNode::release_variables() is called by PyTorch when  the auto gradient graph got destructed (for example, backward run completed or loss function out of its lifecycle). 

Since release_variables is called when some torch Node tears down when the Node's ref cnt being 0 among the overall autograd graph, so we actually don't have a easy way to do what you suggested here.

Reply:
In ORT, PythonOpGrad are responsible to release the shared pointer of corresponding torch context object, which will trigger `release_variables`. But we don't have a way to handle the releasing if backward pass is not called.